### DISCUSSION TITLE
RBE3/Distributing constraint

### DISCUSSION THREAD

**Discussion by b1llg** [idaholab/moose#26571] [Q&A Modules: Solid mechanics]
I would like to implement an RBE3 constraint. I've read about [ConstantRate](https://mooseframework.inl.gov/source/nodalkernels/ConstantRate.html) and I believe it could be a good starting point. However, I need to retrieve all "secondary nodes" position to compute the load applied at each node. 

My goal was to define something like this:
```
[Mesh]
  [rbe3_main]
    type = ExtraNodesetGenerator
    new_boundary = 'main_node'
    coord = '0.0 5.0'
    input = part
  []
[]
[NodalKernels]
  [rbe3_constraint]
    type = DistributingNodalConstraint
    main_node = rbe3_main
    variable = disp_y
    boundary = top
  []
[]
```

First I wanted to know if something similar does already exist ? If not, can I get some suggestion about what should I use and how can I contribute.

Thank you.

**Comment by GiudGiud** [comment]
@hugary1995 and @lindsayad on whether something already exists

**Reply by hugary1995** [reply]
Isn't RBE3 a special case of [`LinearNodalConstraint`](https://mooseframework.inl.gov/source/constraints/LinearNodalConstraint.html)?

**Reply by hugary1995** [reply]
Well okay, I guess if you want to use `LinearNodalConstraint`, you need to set `weights` equal to `1/N` where `N` is the number of primary nodes. This could be tricky if you don't know the number of primary nodes a priori... So maybe creating your own constraint (by deriving from `LinearNodalConstraint`) is the way to go. Anyways, `LinearNodalConstraint` is a good starting point.

**Reply by b1llg** [reply]
It looks like it is doing something similar. The only thing is the weight is not the same on secondary nodes. 
![image](https://github.com/idaholab/moose/assets/65573648/cd734136-3f08-4f63-a638-025841048419)

You can see that in the rbe3, the load applied on the main node is distributed on the selected surface. However, the load applied depends on the position of the secondary nodes vs the main node. So I need to retrieve the nodes on the selected boundaries (or node set in this case?) and compute the weight associated to each node.

**Reply by hugary1995** [reply]
The weights are 1/N.

**Reply by oldninja** [reply]
@hugary1995, your insights into this code are highly valued! I'm interested in understanding the rationale behind setting weights as 1/N in the context of `LinearNodalConstraint`. Could you provide more clarity on the reasons behind this choice? 

It would be beneficial to have a nodal constraint that mimics the function of the RBE3s (Nastran naming) / distributing coupling (Abaqus) directly from the source code.

**Reply by hugary1995** [reply]
I think the corresponding Abaqus theory manual provides a decent explanation of the constraint. https://classes.engineering.wustl.edu/2009/spring/mase5513/abaqus/docs/v6.6/books/stm/default.htm

Without a prescribed rigid body rotation on the master node, each of the slave node is constrained by the translation of the master node. RBE3/distributing coupling satisfies the constraint in an average sense, the weights are therefore uniformly distributed among the slave nodes. In the notation of LinearNodalConstraint that corresponds to weights of 1/N. As stated by the Abaqus user manual, there are alternative ways of distributing the weights, e.g., linear/quadratic decay as a function of master/slave distance.

Neither rigid body rotation nor higher order weights are implemented in LinearNodalConstraint. But as I said, LinearNodalConstraint should be a good place to start. Pull requests are welcome, but a careful benchmarking with either Nastran or Abaqus will be preferred.
